#161348792 Users should receive descriptive signup/Registration validation errors

dnuwa opened the pull request:
#### What does this PR do?
Return descriptive error message to the user
#### Description of Task to be completed?
Have inputs validated on the `POST /api/users/` endpoint ensuring that descriptive error messages are returned to the user
#### How should this be manually tested?
##### To register a user
- Run the application with postman and attempt registering user with incorrect data.
url: api/users
payload:
```
{"user":
	{
		"email":"testuser@text.com",
		"username":"testuser",
		"password":"userpassword"
	
	}	
}
```
##### Error returned
```
{
    "errors": {
        "error": [
            "Ensure your password is alphanumeric, with Minimum eight characters, at least one letter, one number and one special character"
        ]
    }
}
```
#### Any background context you want to provide?
When a user attempts registering with a password that is not alphanumeric and no other special character, the API returns a descriptive error message
This also applies to attempts on submitting an empty password, email, and username fields
#### What are the relevant pivotal tracker stories?
[#161348792](https://www.pivotaltracker.com/story/show/161348792)

emugaya commented on authors/apps/authentication/serializers.py:
```diff
@@ -4,26 +4,51 @@
     # request. Making `token` read-only handles that for us.
 
+    def reg_fields_validation(self, username, email, password):
+        # if email is None:
```
Remove commented code

emugaya replied:
@dnuwa Please respond to this comment

emugaya commented on authors/apps/authentication/serializers.py:
```diff
@@ -4,26 +4,51 @@
+                "Invalid Username , it contains invalid characters."
+            )
+
+        if str(password).isdigit() :
```
use one regex for both alphanumeric e.g `^(?=.*[a-z])(?=.*\d)(?=.*[A-Z])([a-zA-Z\d]{8})$` pattern can take care of your password validations instead of multiple if statements.

dnuwa replied:
Done and it turns out it's very effective.

emugaya commented on authors/apps/authentication/tests/test_model.py:
```diff
@@ -0,0 +1,50 @@
+from rest_framework.test import APITestCase
```
Model testing is good but let's focus on testing endpoints. Please organize tests in folders relating to what your building e.g `auth`

judeinno commented on authors/apps/authentication/serializers.py:
```diff
@@ -4,26 +4,51 @@
         # List all of the fields that could possibly be included in a request
         # or response, including fields specified explicitly above.
-        fields = ['email', 'username', 'password']
+        fields = ["email", "username", "password"]
```
Why are you changing from single quotes to double quotes? Because am using some of those fields and it may lead to merge conflicts.

brucemakallan commented:
We agreed on a tearDown() as a team to avoid creating a new user every time we are testing. I wasn't able to find it in the changed files. Did you find an alternative?

Sulaiman-Mozes commented on authors/apps/authentication/urls.py:
```diff
@@ -6,6 +6,6 @@
 
 urlpatterns = [
```
Its better to provide names for all the urls than just one

dnuwa replied:
@Sulaiman-Mozes kindly color more on this

RachaelNantale commented on authors/apps/authentication/serializers.py:
```diff
@@ -4,26 +4,51 @@
         # List all of the fields that could possibly be included in a request
         # or response, including fields specified explicitly above.
-        fields = ['email', 'username', 'password']
+        fields = ["email", "username", "password"]
```
Why did you change to double quotes?

dnuwa replied:
@RachaelNantale this was done by pep8 formatting. since we resolved to use single quotes, this file has been updated to have single quotes.

emugaya commented:
@dnuwa Please follow conventions in the engineering playbook on how to write PRs

emugaya commented on authors/apps/authentication/validators.py:
```diff
@@ -0,0 +1,34 @@
+            raise APIException(
+                {'username_error':'Invalid Username , it contains invalid characters.'}
+            )
+        
```
Add newline

Commit:
feature(descriptive signup error messages): Users should receive descriptive signup/Registration validation errors
- existing user email validation
- invalid email validation
- password of length less than 8 validation
- add password not being alphanumeric validation
- add tests for registering user validations
- add tests for creating user models
[Fixes #161348792]

emugaya reviewed (approved):
Fix Heroku review App before I merge

emugaya commented on authors/apps/authentication/tests_auth/test_model.py:
```diff
@@ -0,0 +1,50 @@
+        self.assertEqual(self.user.email, "test@test.com")
+
+    def test_missing_username(self):
+        try:
```
Why do you have `try catch` blocks in tests?

emugaya reviewed (changes requested):
Your model tests are lacking. Read on how to write `django` model tests. I advised against writing them unless your sure of what you are doing.
